File: benchmark.py
# FIBONACCI HISTORICAL RUNNER
# BUILD: exact-target-v3.0 / 2026-09-01

from __future__ import annotations
import ast, hashlib, math, os, statistics, subprocess, time
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

BUILD = "exact-target-v3.0 / 2026-09-01"
ROOT = Path(__file__).resolve().parent
V1_FILE, V2_FILE, V3_FILE = ROOT/'fibv1.py', ROOT/'fibv2.py', ROOT/'fibv3.py'
V4_EXE = ROOT / ('fibv4.exe' if os.name == 'nt' else 'fibv4')
Timer = Callable[[int, int], tuple[float, list[float]]]

@dataclass
class Candidate:
    n: int
    median: float
    samples: list[float]
    def error(self, target: float) -> float:
        return abs(self.median-target)/target
    def percent(self, target: float) -> float:
        return self.median/target*100.0

def sha12(path): return hashlib.sha256(path.read_bytes()).hexdigest()[:12]

def load_function_only(path: Path, name: str):
    tree = ast.parse(path.read_text(encoding='utf-8'), filename=str(path))
    keep = [n for n in tree.body if isinstance(n,(ast.FunctionDef,ast.AsyncFunctionDef,ast.Import,ast.ImportFrom))]
    mod = ast.Module(body=keep, type_ignores=[]); ast.fix_missing_locations(mod)
    ns={'__name__':'_fib_benchmark_loaded_'}
    exec(compile(mod,str(path),'exec'),ns,ns)
    f=ns.get(name)
    if not callable(f): raise RuntimeError(f'{name} not found in {path.name}')
    return f

def python_timer(func):
    def timer(n,trials):
        xs=[]
        for _ in range(max(1,trials)):
            s=time.perf_counter_ns(); r=func(n); e=time.perf_counter_ns()
            if r is None: raise RuntimeError('Fibonacci function returned None')
            xs.append((e-s)/1e9)
        return statistics.median(xs),xs
    return timer

def cpp_timer(n,trials):
    xs=[]
    for _ in range(max(1,trials)):
        cp=subprocess.run([str(V4_EXE),'--bench',str(n),'1'],cwd=ROOT,capture_output=True,text=True,check=True)
        out=cp.stdout.strip()
        if not out: raise RuntimeError('fibv4.exe produced no timing')
        xs.append(float(out.splitlines()[-1]))
    return statistics.median(xs),xs

def measure(timer,n,trials):
    med,xs=timer(max(0,int(n)),trials); return Candidate(max(0,int(n)),med,xs)

def better(a,b,target):
    if a is None or b.error(target)<a.error(target): return b
    if abs(b.error(target)-a.error(target))<1e-12 and b.n>a.n: return b
    return a

def warmup(timer):
    try: timer(1000,1)
    except Exception: pass

def rough_bracket(timer,target,start=1024):
    n=max(2,start); low=Candidate(0,0.0,[0.0]); visited=[low]; runs=0
    while True:
        c=measure(timer,n,1); runs+=1; visited.append(c)
        if c.median<target:
            low=c; n*=2; continue
        c=measure(timer,n,3); runs+=3; visited.append(c)
        if c.median>=target: return low,c,visited,runs
        low=c; n*=2

def interpolated_n(low,high,target):
    lo,hi=low.n,high.n
    if hi-lo<=2: return lo+(hi-lo)//2
    dt=high.median-low.median
    if dt<=1e-12: g=(lo+hi)//2
    else:
        f=(target-low.median)/dt; f=min(.85,max(.15,f)); g=lo+int((hi-lo)*f)
    return min(hi-1,max(lo+1,g))

def classify(err):
    p=err*100
    return 'EXCELLENT' if p<=.10 else 'GREAT' if p<=.25 else 'GOOD' if p<=.50 else 'CLOSE' if p<=1 else 'NOISY'

def search_fast(timer,user_target):
    desired=user_target*.90
    low,high,seen,runs=rough_bracket(timer,desired)
    best=None
    for c in seen: best=better(best,c,desired)
    for _ in range(7):
        if high.n-low.n<=2: break
        c=measure(timer,interpolated_n(low,high,desired),1); runs+=1; best=better(best,c,desired)
        if c.median<desired: low=c
        else: high=c
    best=measure(timer,best.n,3); runs+=3
    pct=best.median/user_target
    if pct<.85 or pct>.95:
        rn=max(1,int(best.n*desired/max(best.median,1e-12)))
        r=measure(timer,rn,3); runs+=3
        if abs(r.median-desired)<abs(best.median-desired): best=r
    return best,runs

def search_accurate(timer,target):
    """
    STUBBORN accurate mode.

    Keep moving n higher/lower until we are genuinely very close to the
    user's requested runtime. 98% is NOT a finish line.

    Stop conditions:
      1) certified median is within 0.10% of target, OR
      2) the integer n bracket collapses, OR
      3) we exhaust a large search budget.

    We always remember the closest median ever seen, whether it is
    slightly UNDER or slightly OVER the target.
    """
    low,high,seen,runs=rough_bracket(timer,target)
    best=None
    cand={}

    for c in seen:
        cand[c.n]=c
        best=better(best,c,target)

    # Search much more stubbornly than v3.
    # Trial count increases automatically as we approach the bullseye.
    for _ in range(40):
        if high.n-low.n<=1:
            break

        current_error=best.error(target)

        if current_error > .02:       # >2% away
            trials=3
        elif current_error > .005:    # 0.5%-2%
            trials=5
        else:                         # already close
            trials=7

        n=interpolated_n(low,high,target)

        # Avoid repeatedly testing the same point when interpolation rounds.
        if n in cand and high.n-low.n>2:
            n=(low.n+high.n)//2
            if n in cand:
                n=min(high.n-1,max(low.n+1,n+1))

        c=measure(timer,n,trials)
        runs+=trials
        cand[n]=c
        best=better(best,c,target)

        if c.median<target:
            low=c
        else:
            high=c

        # Proportional bullseye shot from the best measurement so far.
        # This is especially useful when we are sitting at 97-99%.
        if best.error(target) <= .03 and high.n-low.n>1:
            ratio=target/max(best.median,1e-12)
            bn=max(low.n+1,min(high.n-1,int(round(best.n*ratio))))

            if bn not in cand and low.n < bn < high.n:
                btrials=5 if best.error(target)>.005 else 7
                b=measure(timer,bn,btrials)
                runs+=btrials
                cand[bn]=b
                best=better(best,b,target)

                if b.median<target:
                    low=b
                else:
                    high=b

        # Once search measurements themselves are very close, certify now.
        if best.error(target) <= .001:   # <=0.10%
            cert=measure(timer,best.n,9)
            runs+=9
            cand[best.n]=cert
            best=better(best,cert,target)
            if cert.error(target) <= .001:
                return cert,runs

    # Search is over or bracket collapsed.
    # Certify several of the strongest candidates with heavier sampling.
    ranked=sorted(cand,key=lambda n:cand[n].error(target))
    final=None

    for n in ranked[:5]:
        c=measure(timer,n,9)
        runs+=9
        final=better(final,c,target)

        if c.error(target)<=.001:
            # One stronger confirmation before accepting the bullseye.
            confirm=measure(timer,n,11)
            runs+=11
            final=better(final,confirm,target)
            if confirm.error(target)<=.001:
                return confirm,runs

    # If certification knocked us away from target, keep correcting.
    # Do NOT quit at 98%. Up to 10 extra feedback rounds.
    for _ in range(10):
        if final.error(target)<=.001:
            break

        ratio=target/max(final.median,1e-12)
        n=max(1,int(round(final.n*ratio)))

        if n==final.n:
            n=max(1,n+(1 if final.median<target else -1))

        c=measure(timer,n,7)
        runs+=7
        final=better(final,c,target)

        if c.error(target)<=.001:
            confirm=measure(timer,c.n,11)
            runs+=11
            final=better(final,confirm,target)
            if confirm.error(target)<=.001:
                return confirm,runs

    # Final strongest measurement. This can still miss 0.10% if the machine's
    # timing noise is larger than the requested precision, but it will return
    # the closest result we actually measured rather than giving up at 98%.
    winner=measure(timer,final.n,13)
    runs+=13

    if winner.error(target) <= final.error(target):
        return winner,runs

    # Final's earlier median was closer; re-certify it once so the returned
    # result always corresponds to a fresh heavy sample set.
    retry=measure(timer,final.n,13)
    runs+=13
    return retry,runs

def fmt_range(xs): return '-' if not xs else f'{min(xs):.6f}..{max(xs):.6f}'

def print_result(name,c,target,runs,wall,fast=False):
    pct=c.percent(target); err=c.error(target)*100
    label=('OK' if 85<=pct<=95 else '~') if fast else classify(c.error(target))
    side='UNDER' if c.median<target else 'OVER' if c.median>target else 'EXACT'
    print(f'  {name:<29} F_{c.n:,}')
    print(f'      median {c.median:.6f} s | {pct:7.3f}% target | error {err:.3f}% | {side} | {label}')
    print(f'      range {fmt_range(c.samples)} | runs {runs} | wall {wall:.2f}s')

def preflight():
    for p in (V1_FILE,V2_FILE,V3_FILE):
        if not p.exists(): raise FileNotFoundError(f'Missing {p.name}')
    if not V4_EXE.exists(): raise FileNotFoundError(f'Missing {V4_EXE.name}. Compile fibv4.cpp first.')
    f1=load_function_only(V1_FILE,'fib_v1'); f2=load_function_only(V2_FILE,'fib_pair'); f3=load_function_only(V3_FILE,'fib_fast')
    if f1(20)!=6765 or f2(20)[0]!=6765 or f3(20)!=6765: raise RuntimeError('Python correctness preflight failed')
    cpp_timer(1000,1)
    return [('V1 Python iterative',python_timer(f1)),('V2 Python recursive doubling',python_timer(f2)),('V3 Python iterative doubling',python_timer(f3)),('V4 C++/GMP doubling',cpp_timer)]

def run_mode(versions,mode,target):
    if mode=='fast': print(f'\nFAST MODE — aiming for ~90% of user target ({target*.90:.6f} s)')
    else:
        print(f'\nACCURATE MODE — chasing exact requested target ({target:.6f} s)')
        print('Slightly UNDER and slightly OVER are equally valid; absolute error wins.')
    for name,timer in versions:
        print(f'\nTesting {name}...'); warmup(timer); w=time.perf_counter()
        c,runs=(search_fast(timer,target) if mode=='fast' else search_accurate(timer,target))
        print_result(name,c,target,runs,time.perf_counter()-w,fast=(mode=='fast'))

def main():
    print('FIBONACCI HISTORICAL BENCHMARK'); print(f'Runner build : {BUILD}\n')
    print('1. Fast     - aims around 90% of requested runtime')
    print('2. Accurate - chases requested runtime as closely as possible')
    print('3. Both\n')
    choice=input('Choose mode [1]: ').strip() or '1'
    try: target=float(input('Target seconds [1.0]: ').strip() or '1.0')
    except ValueError: print('Invalid target.'); return 1
    if target<=0: print('Target must be > 0.'); return 1
    try: versions=preflight()
    except Exception as e: print(f'\nPREFLIGHT ERROR: {e}'); return 1
    print(f'\nLoaded V1 : {V1_FILE.name} sha256 {sha12(V1_FILE)}')
    print(f'Loaded V2 : {V2_FILE.name} sha256 {sha12(V2_FILE)}')
    print(f'Loaded V3 : {V3_FILE.name} sha256 {sha12(V3_FILE)}')
    print(f'Loaded V4 : {V4_EXE.name}\nPreflight : OK')
    print(f'\nUser target : {target:.6f} s')
    print('V1-V3 clock: time.perf_counter_ns() around function call only')
    print('V4 clock   : std::chrono::steady_clock inside fibv4.exe')
    print('Controller : chooses n between timed runs; controller overhead is not score time')
    total=time.perf_counter()
    if choice=='1': run_mode(versions,'fast',target)
    elif choice=='2': run_mode(versions,'accurate',target)
    elif choice=='3': run_mode(versions,'fast',target); run_mode(versions,'accurate',target)
    else: print('Invalid mode.'); return 1
    print('\n'+'='*76); print(f'TOTAL BENCHMARK WALL TIME: {time.perf_counter()-total:.3f} s'); print('='*76)
    return 0

if __name__=='__main__': raise SystemExit(main())

File: benchmark.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <limits>
#include <string>
#include <utility>
#include <vector>

#include <gmp.h>
#include <gmpxx.h>

#ifdef _WIN32
#define NOMINMAX
#include <windows.h>
#else
#include <time.h>
#endif

using u64 = std::uint64_t;
using FibFn = mpz_class (*)(u64);

static volatile unsigned long g_sink = 0;
static void consume(const mpz_class& x) {
    g_sink ^= mpz_get_ui(x.get_mpz_t());
}

// ============================================================
// Fibonacci implementations
// ============================================================

mpz_class fib_v1(u64 n) {
    mpz_class a = 0, b = 1;
    for (u64 i = 0; i < n; ++i) {
        mpz_class next = a + b;
        a = std::move(b);
        b = std::move(next);
    }
    return a;
}

std::pair<mpz_class, mpz_class> fib_pair_v2(u64 n) {
    if (n == 0) return {0, 1};

    auto p = fib_pair_v2(n >> 1);
    const mpz_class& a = p.first;
    const mpz_class& b = p.second;

    mpz_class c = a * ((b << 1) - a);
    mpz_class d = a * a + b * b;

    if ((n & 1ULL) == 0) return {std::move(c), std::move(d)};
    return {d, c + d};
}

mpz_class fib_v2(u64 n) {
    return fib_pair_v2(n).first;
}

mpz_class fib_v3(u64 n) {
    if (n == 0) return 0;

    mpz_class a = 0, b = 1;
    int highest = 63;
    while (highest > 0 && ((n >> highest) & 1ULL) == 0) --highest;

    for (int bit = highest; bit >= 0; --bit) {
        mpz_class c = a * ((b << 1) - a);
        mpz_class d = a * a + b * b;

        if (((n >> bit) & 1ULL) == 0) {
            a = std::move(c);
            b = std::move(d);
        } else {
            mpz_class e = c + d;
            a = std::move(d);
            b = std::move(e);
        }
    }
    return a;
}

// Kept separate so V4 can be replaced later without touching the runner.
std::pair<mpz_class, mpz_class> fib_pair_v4(u64 n) {
    if (n == 0) return {0, 1};

    auto p = fib_pair_v4(n >> 1);
    mpz_class a = std::move(p.first);
    mpz_class b = std::move(p.second);

    mpz_class c = a * ((b << 1) - a);
    mpz_class d = a * a + b * b;

    if ((n & 1ULL) == 0) return {std::move(c), std::move(d)};
    return {d, c + d};
}

mpz_class fib_v4(u64 n) {
    return fib_pair_v4(n).first;
}

struct Impl {
    const char* name;
    FibFn fn;
    // Approximate local exponent t ~ n^p, used only as a conservative controller.
    // We do NOT fit p from noisy near-boundary timings.
    double p_hint;
};

static const Impl IMPLS[] = {
    {"V1 iterative",            fib_v1, 1.80},
    {"V2 recursive doubling",  fib_v2, 1.30},
    {"V3 iterative doubling",  fib_v3, 1.30},
    {"V4 GMP doubling",        fib_v4, 1.30},
};

// ============================================================
// Timing
// ============================================================

// Current-thread CPU time is the search metric. It excludes time where Windows
// deschedules the benchmark, which was the biggest source of the old 90%-110%
// swings. Wall time is still measured and printed beside it.
double cpu_now() {
#ifdef _WIN32
    FILETIME creation{}, exit{}, kernel{}, user{};
    if (!GetThreadTimes(GetCurrentThread(), &creation, &exit, &kernel, &user)) {
        return 0.0;
    }

    ULARGE_INTEGER k{}, u{};
    k.LowPart = kernel.dwLowDateTime;
    k.HighPart = kernel.dwHighDateTime;
    u.LowPart = user.dwLowDateTime;
    u.HighPart = user.dwHighDateTime;

    return static_cast<double>(k.QuadPart + u.QuadPart) * 1e-7;
#else
    timespec ts{};
    if (clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts) != 0) return 0.0;
    return static_cast<double>(ts.tv_sec) + static_cast<double>(ts.tv_nsec) * 1e-9;
#endif
}

struct OneSample {
    double cpu = 0.0;
    double wall = 0.0;
};

OneSample time_once(FibFn fn, u64 n) {
    auto w0 = std::chrono::steady_clock::now();
    double c0 = cpu_now();

    mpz_class result = fn(n);

    double c1 = cpu_now();
    auto w1 = std::chrono::steady_clock::now();
    consume(result);

    return {
        c1 - c0,
        std::chrono::duration<double>(w1 - w0).count()
    };
}

double median(std::vector<double> v) {
    std::sort(v.begin(), v.end());
    const std::size_t m = v.size() / 2;
    if (v.size() & 1U) return v[m];
    return 0.5 * (v[m - 1] + v[m]);
}

struct Measurement {
    double cpu = 0.0;
    double wall = 0.0;
    double min_cpu = 0.0;
    double max_cpu = 0.0;
    int samples = 0;
};

Measurement measure(FibFn fn, u64 n, int samples) {
    std::vector<double> cpus;
    std::vector<double> walls;
    cpus.reserve(samples);
    walls.reserve(samples);

    for (int i = 0; i < samples; ++i) {
        OneSample s = time_once(fn, n);
        cpus.push_back(s.cpu);
        walls.push_back(s.wall);
    }

    Measurement m;
    m.cpu = median(cpus);
    m.wall = median(walls);
    m.min_cpu = *std::min_element(cpus.begin(), cpus.end());
    m.max_cpu = *std::max_element(cpus.begin(), cpus.end());
    m.samples = samples;
    return m;
}

// ============================================================
// Controller helpers
// ============================================================

u64 scale_n(u64 n, long double factor) {
    if (n == 0) n = 1;

    long double x = static_cast<long double>(n) * factor;
    if (x < 1.0L) return 1;
    if (x >= static_cast<long double>(std::numeric_limits<u64>::max())) {
        return std::numeric_limits<u64>::max();
    }

    u64 out = static_cast<u64>(x);
    if (out == n) {
        if (factor > 1.0L && n < std::numeric_limits<u64>::max()) ++out;
        else if (factor < 1.0L && n > 1) --out;
    }
    return out;
}

void warmup(FibFn fn) {
    consume(fn(4096));
    consume(fn(4352));
}

// Cheaply get to roughly the right order of magnitude.
u64 coarse_seed(FibFn fn, double goal, double p_hint) {
    u64 n = 1024;
    double t = 0.0;

    for (int i = 0; i < 64; ++i) {
        t = time_once(fn, n).cpu;
        if (t >= goal * 0.20) break;

        if (n > std::numeric_limits<u64>::max() / 2) break;
        n *= 2;
    }

    if (t > 0.0) {
        long double factor = std::pow(
            static_cast<long double>(goal / t),
            1.0L / static_cast<long double>(p_hint)
        );
        factor = std::clamp(factor, 0.75L, 3.00L);
        n = scale_n(n, factor);
    }

    return n;
}

struct SearchResult {
    u64 n = 0;
    Measurement m;
    bool in_band = false;
};

// ============================================================
// FAST: aim around 90% of target
// ============================================================

SearchResult run_fast_one(const Impl& impl, double target) {
    const double goal = 0.90 * target;
    const double low = 0.86 * target;
    const double high = 0.94 * target;

    warmup(impl.fn);
    u64 n = coarse_seed(impl.fn, goal, impl.p_hint);

    // Cheap one-sample steering.
    for (int round = 0; round < 5; ++round) {
        double t = time_once(impl.fn, n).cpu;
        if (t >= low && t <= high) break;

        long double factor = std::pow(
            static_cast<long double>(goal / std::max(t, 1e-12)),
            1.0L / static_cast<long double>(impl.p_hint)
        );
        factor = std::clamp(factor, 0.78L, 1.28L);
        n = scale_n(n, factor);
    }

    Measurement m = measure(impl.fn, n, 3);

    // At most two corrections. Fast should stay fast.
    for (int round = 0; round < 4 && (m.cpu < low || m.cpu > high); ++round) {
        long double factor = std::pow(
            static_cast<long double>(goal / std::max(m.cpu, 1e-12)),
            1.0L / static_cast<long double>(impl.p_hint)
        );
        factor = std::clamp(factor, 0.84L, 1.16L);
        if (m.cpu > target) factor = std::min<long double>(factor, 0.92L);
        else if (m.cpu > high) factor = std::min<long double>(factor, 0.97L);
        else if (m.cpu < low) factor = std::max<long double>(factor, 1.03L);
        n = scale_n(n, factor);
        m = measure(impl.fn, n, 3);
    }

    return {n, m, m.cpu >= low && m.cpu <= high};
}

// ============================================================
// ACCURATE: certify a 99%-100% median
// ============================================================

SearchResult run_accurate_one(const Impl& impl, double target) {
    const double goal = 0.995 * target;
    const double low = 0.990 * target;
    const double high = 1.000 * target;

    warmup(impl.fn);
    u64 n = coarse_seed(impl.fn, goal, impl.p_hint);

    u64 best_n = 0;
    Measurement best_m{};

    auto remember_pass = [&](u64 candidate_n, const Measurement& m) {
        if (m.cpu <= target &&
            (best_n == 0 || m.cpu > best_m.cpu)) {
            best_n = candidate_n;
            best_m = m;
        }
    };

    // Stage A: median-of-3 feedback. This gets very close without spending
    // dozens of 1-second samples on binary-search points we do not care about.
    for (int round = 0; round < 9; ++round) {
        Measurement m = measure(impl.fn, n, 3);
        remember_pass(n, m);

        if (m.cpu >= low && m.cpu <= high) {
            Measurement verify = measure(impl.fn, n, 7);
            remember_pass(n, verify);
            if (verify.cpu >= low && verify.cpu <= high) {
                return {n, verify, true};
            }
            m = verify;
        }

        long double factor = std::pow(
            static_cast<long double>(goal / std::max(m.cpu, 1e-12)),
            1.0L / static_cast<long double>(impl.p_hint)
        );

        // Damped control. Close to the target, never jump by more than 3% in n.
        const double error = std::abs(m.cpu - goal) / target;
        if (error < 0.06) factor = std::clamp(factor, 0.970L, 1.030L);
        else              factor = std::clamp(factor, 0.850L, 1.150L);

        n = scale_n(n, factor);
    }

    // Stage B: high-confidence feedback with median-of-7.
    for (int round = 0; round < 6; ++round) {
        Measurement m = measure(impl.fn, n, 7);
        remember_pass(n, m);

        if (m.cpu >= low && m.cpu <= high) {
            return {n, m, true};
        }

        long double factor = std::pow(
            static_cast<long double>(goal / std::max(m.cpu, 1e-12)),
            1.0L / static_cast<long double>(impl.p_hint)
        );
        factor = std::clamp(factor, 0.975L, 1.025L);
        n = scale_n(n, factor);
    }

    // Stage C: final median-of-9 certification. If drift moved the candidate,
    // feed it back one last time rather than accepting an OVER result.
    for (int round = 0; round < 4; ++round) {
        Measurement m = measure(impl.fn, n, 9);
        remember_pass(n, m);

        if (m.cpu >= low && m.cpu <= high) {
            return {n, m, true};
        }

        long double factor = std::pow(
            static_cast<long double>(goal / std::max(m.cpu, 1e-12)),
            1.0L / static_cast<long double>(impl.p_hint)
        );
        factor = std::clamp(factor, 0.980L, 1.020L);
        n = scale_n(n, factor);
    }

    // Never lie. Re-check the closest under-target result; if it drifted OVER,
    // back off until a fresh 9-sample median is <= target.
    if (best_n != 0) {
        n = best_n;
        Measurement check = measure(impl.fn, n, 9);
        if (check.cpu >= low && check.cpu <= target) {
            return {n, check, true};
        }

        // If the best pass drifted slightly low, actively walk it back toward 99.5%.
        // This is intentionally slow/strong: Accurate mode is allowed to spend time.
        if (check.cpu < low) {
            for (int round = 0; round < 6; ++round) {
                long double factor = std::pow(
                    static_cast<long double>(goal / std::max(check.cpu, 1e-12)),
                    1.0L / static_cast<long double>(impl.p_hint)
                );
                factor = std::clamp(factor, 1.002L, 1.020L);
                u64 next = scale_n(n, factor);
                Measurement trial = measure(impl.fn, next, 9);

                if (trial.cpu >= low && trial.cpu <= target) {
                    return {next, trial, true};
                }

                if (trial.cpu > target) {
                    // Split the index interval instead of accepting OVER.
                    next = n + (next - n) / 2;
                    trial = measure(impl.fn, next, 9);
                    if (trial.cpu >= low && trial.cpu <= target) {
                        return {next, trial, true};
                    }
                    if (trial.cpu > target) {
                        // Keep n as the confirmed lower candidate.
                        check = measure(impl.fn, n, 9);
                        continue;
                    }
                }

                n = next;
                check = trial;
            }
        }

        if (check.cpu <= target && check.cpu >= low) {
            return {n, check, true};
        }
    }

    for (int round = 0; round < 8; ++round) {
        n = scale_n(n, 0.99L);
        Measurement m = measure(impl.fn, n, 9);
        if (m.cpu <= target) {
            return {n, m, m.cpu >= low};
        }
    }

    return {best_n, best_m, false};
}

// ============================================================
// Output / CLI
// ============================================================

void print_result(const Impl& impl, const SearchResult& r, double target) {
    double pct = (r.m.cpu / target) * 100.0;

    std::cout
        << std::left << std::setw(24) << impl.name
        << "F_" << std::setw(13) << r.n
        << " CPU " << std::fixed << std::setprecision(6) << r.m.cpu << " s"
        << "  " << std::setw(7) << std::setprecision(2) << pct << "%"
        << " wall " << std::setprecision(6) << r.m.wall << " s"
        << "  range " << r.m.min_cpu << ".." << r.m.max_cpu
        << "  " << (r.in_band ? "OK" : "UNSTABLE")
        << '\n';
}

void run_fast(double target) {
    std::cout << "\nFAST  (goal: about 90% of target)\n";
    for (const auto& impl : IMPLS) {
        print_result(impl, run_fast_one(impl, target), target);
    }
}

void run_accurate(double target) {
    std::cout << "\nACCURATE  (certified median goal: 99%-100% of target)\n";
    for (const auto& impl : IMPLS) {
        print_result(impl, run_accurate_one(impl, target), target);
    }
}

int main(int argc, char** argv) {
    std::string mode;
    double target = 1.0;

    if (argc >= 2) mode = argv[1];
    if (argc >= 3) {
        try { target = std::stod(argv[2]); }
        catch (...) { std::cerr << "Invalid target.\n"; return 1; }
    }

    if (mode != "fast" && mode != "accurate" && mode != "both") {
        std::cout
            << "FIBONACCI BENCHMARK RUNNER\n\n"
            << "Search metric: current-thread CPU time\n"
            << "Wall time is reported beside it.\n\n"
            << "1. Fast      (~90% target)\n"
            << "2. Accurate  (99%-100% target)\n"
            << "3. Both\n\n"
            << "Choice [1]: ";

        std::string choice;
        std::getline(std::cin, choice);
        if (choice.empty() || choice == "1") mode = "fast";
        else if (choice == "2") mode = "accurate";
        else if (choice == "3") mode = "both";
        else { std::cerr << "Invalid mode.\n"; return 1; }

        std::cout << "Target seconds [1.0]: ";
        std::string s;
        std::getline(std::cin, s);
        if (!s.empty()) {
            try { target = std::stod(s); }
            catch (...) { std::cerr << "Invalid target.\n"; return 1; }
        }
    }

    if (target <= 0.0) {
        std::cerr << "Target must be > 0.\n";
        return 1;
    }

    std::cout
        << "\nTarget: " << target << " s\n"
        << "Search metric: thread CPU time (scheduler pauses excluded)\n";

    if (mode == "fast" || mode == "both") run_fast(target);
    if (mode == "accurate" || mode == "both") run_accurate(target);

    if (g_sink == std::numeric_limits<unsigned long>::max()) {
        std::cerr << "sink=" << g_sink << '\n';
    }

    return 0;
}
